crates/common/src/commit/request.rs

```diff
@@ -32,24 +62,60 @@ impl SignedProxyDelegation {
     }
 }
 
+impl<T: PublicKey> fmt::Display for SignedProxyDelegation<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}\nSignature: {}", self.message, self.signature)
+    }
+}
+
+// TODO(David): This struct shouldn't be visible to module authors
 #[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct SignRequest {
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum SignRequest {
+    Consensus(SignConsensusRequest),
+    ProxyBls(SignProxyRequest<BlsPublicKey>),
+    ProxyEcdsa(SignProxyRequest<EcdsaPublicKey>),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignConsensusRequest {
     pub pubkey: BlsPublicKey,
-    pub is_proxy: bool,
+    #[serde(with = "alloy::hex::serde")]
     pub object_root: [u8; 32],
 }
 
-impl SignRequest {
-    pub fn new(pubkey: BlsPublicKey, is_proxy: bool, object_root: [u8; 32]) -> SignRequest {
-        Self { pubkey, is_proxy, object_root }
+impl SignConsensusRequest {
```

Comment:
I wonder if we can do something like:

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SignRequestType<T: PublicKey> {
    pub pubkey: T,
    #[serde(with = "alloy::hex::serde")]
    pub object_root: [u8; 32],
}

impl<T: PublicKey> SignRequestType<T> {
    pub fn new(pubkey: T, object_root: [u8; 32]) -> Self {
        Self { pubkey, object_root }
    }

    pub fn builder(pubkey: T) -> Self {
        Self::new(pubkey, [0; 32])
    }

    pub fn with_root(self, object_root: [u8; 32]) -> Self {
        Self { object_root, ..self }
    }

    pub fn with_msg(self, msg: &impl TreeHash) -> Self {
        self.with_root(msg.tree_hash_root().0)
    }
}

// Example usage for BlsPublicKey:
pub type SignConsensusRequest = SignRequestType<BlsPublicKey>;

// Example usage for other types of PublicKey:
pub type SignProxyRequest<T> = SignRequestType<T>;
```
to reduce code duplication

Reply:
I actually had exactly this going on, but then decided to explicitly differentiate the request types for a consensus request and for a bls proxy request as we want the user-facing API to be as straightforward as possible. A type alias can be considered helpful enough, but we lose the type guarantee we now have after the use has explicitly constructed a request type called `SignConsensusRequest` 😄 So with the suggested type aliasing, technically a `SignProxyRequest<BlsPublicKey>` could still be passed as an argument to `request_consensus_signature`.

With this in mind, I'd still say the code duplication is worth reducing, so if you deem the little drawback outlined above is worth sacrificing, I'm on board with that 👍

Reply:
It's a valid point, let's keep for now, eventually we will change afterwards